proof-language.md

```diff
@@ -0,0 +1,359 @@
+-   Easily verifiable using ZK
+-   Low memory usage.
+-   Composability
+
+
+## Non-goals:
+
+-   Parallization: I have not been thinking explicitly about parallel verification of this format,
+    because I expect parallization to happen by composing units of this format into larger proofs.
+    Even so I think with some additional metadata, we could make
+    this format parallizable.
+
+-   Human readability: We aim to be a simple binary format. To become human
+    readable, we expect the user to run a secondary program to pretty-print the
+    proof. Note that symbols, lemmas, and notation do not have names.
+
+
+Terms
+=====
+
+As mentioned previously, we may view this format as a language for constructing
+terms that meet particular well-formedness criteria. Let us first enumerate
+those terms, in a python-esque pseudocode.
+
+`Term`s are of two types, `Pattern`, and `Proof`.
+Every `Term` has a well-formedness check.
+Every `Proof` has a conclusion.
+
+```python
+abstract class Term:
+    abstract def well_formed:
+        ...
+
+abstract class Pattern(Term):
+    def well_formed():
+        # Fails if any two MetaVars with the same id has
+        # different meta-requirements
+        # MetaVars are introduced later.
+    ...
+
+abstract class Proof:
+    abstract def conclusion:
+        ...
+```
+
+`Pattern`s
+----------
+
+There are two classes of `Pattern`,
+the expected concrete constructors for matching logic patterns,
+and a representation for various "meta" patterns.
+
+### Standard patterns
+
+```python
+class Symbol(Pattern)
+    name: uint32
+
+class SVar(Pattern)
+    name: uint32
+
+class EVar(Pattern)
+    name: uint32
+
+class Implication(Pattern)
+    left: Pattern
+    right: Pattern
+
+class Application(Pattern)
+    left: Pattern
+    right: Pattern
+
+class Exists(Pattern)
+    var: EVar
+    subpattern : Pattern
+
+class Mu(Pattern)
+    var: SVar
+    subpattern : Pattern
+
+    def well_formed():
+        return super.well_formed()
+           and subpattern.var_occurs_positively(var)
+```
+
+### Meta-patterns
+
+Meta-patterns allow us to represent axiom- and theorem-schemas through the use
+of metavariables:
+
+```python
+class MetaVar : Pattern
+    name: uint32
+
+    # Meta-requirements that must be satisfied by any instantiation.
+    fresh: list[EVar | SVar]        # variables that must not occur in an instatiation
+    positive: list[SVar]            # variables that must only occur positively in an instatiation
+    negative: list[SVar]            # variables that must only occur negatively in an instatiation
+    application_context: list[EVar] # Variables that must only occur as a hole variable in an application context.
+```
+
+Each `MetaVar` has a list of requirements that must be met by any instantiation.
+These may also be used by the well-formedness checks for `Proof`s.
+
+We also need to represent substitutions applied to `MetaVar`s.
+
+```python
+class ESubst : Pattern
```

Comment:
Isn't it still `ESubst : MetaVar`? You are substituting in an unspecified pattern.

This would also make sense for `InstantiateSchema` : if it comes across `ESubst`, it can perform the substitution right after the instantiation.

Reply:
Sorry, my syntax is all over the place. This should be `class ESubst(Pattern):`. Every `ESubst` is a pattern, but contains a metavariable.

Reply:
Maybe I'm not getting it, but I don't think ESubst should be a pattern if it is meant to represent substitutions to metavars. It should be a MetaVar. Pattern is meant to represent concrete patterns.

Let's say you have MetaVar phi.
And then you do phi[psi/x].
Since phi is still unknown, phi[psi/x] is still in a way MetaVar. You cannot do a well formedness check on such a thing, hence it's not a term.

Reply:
Hm now i see that metavar is also in Pattern. I'm not sure if that makes sense.

Reply:
I'm not sure I understand. Are there any cases when a substitution is ill-formed? Given that `psi` is well formed, why under what conditions is `phi[psi/x]` ill-formed?

Note that `matching-logic.mm` also considers metavariables patterns when declared with a `$f` statement.

Reply:
Note that the checking of freshness, positivity, etc of `phi[psi/x]` will have to be best effort, and fail when we can't prove things conclusively. 

The obvious ones we want are:

1. if `y` is fresh in `phi`  and `psi`, then `y` is fresh in `phi[psi/x]`.
2. if `x` is positive in `phi` and `y` is positive is `psi`, then `y` is positive in `phi[psi/x]`. (resp for negative). 
3. if `phi` is an application context for `x` and `psi` is an application context for `y`, then `phi[psi/x]` is an application context for `y`.

Reply:
It would be interesting if we could have meta completeness results for well-formedness checks. Perhaps we should bring this up in a discussion on the `ml-formalization` channel.